File: src/strats_prob/exits.py
"""Exit management for the probe engine.

Handles ATR-based stops, targets, trailing stops, and time stops
with risk-profile scaling.
"""

import logging
from dataclasses import dataclass
from typing import Optional

logger = logging.getLogger(__name__)


@dataclass
class RiskProfile:
    """Risk profile that scales exit parameters."""

    name: str
    stop_scale: float
    target_scale: float
    trail_scale: float
    time_scale: float


RISK_PROFILES: dict[str, RiskProfile] = {
    "low": RiskProfile(name="low", stop_scale=0.75, target_scale=0.75, trail_scale=0.75, time_scale=0.6),
    "medium": RiskProfile(name="medium", stop_scale=1.0, target_scale=1.0, trail_scale=1.0, time_scale=1.0),
    "high": RiskProfile(name="high", stop_scale=1.5, target_scale=1.5, trail_scale=1.5, time_scale=1.5),
}


class ExitManager:
    """Manages exit logic for a single trade.

    Initialized at entry with risk-profile-scaled parameters.
    Call check() each bar to determine if the trade should be closed.
    Tracks max favorable excursion (MFE) and max adverse excursion (MAE).
    """

    def __init__(
        self,
        entry_price: float,
        direction: str,
        atr_at_entry: float,
        atr_stop_mult: Optional[float],
        atr_target_mult: Optional[float],
        trailing_atr_mult: Optional[float],
        time_stop_bars: Optional[int],
        risk_profile: RiskProfile,
    ):
        self.entry_price = entry_price
        self.direction = direction  # "long" or "short"
        self.atr = atr_at_entry
        self.risk_profile = risk_profile

        # Scale multipliers by risk profile
        self.stop_dist = (atr_stop_mult * risk_profile.stop_scale * self.atr) if atr_stop_mult else None
        self.target_dist = (atr_target_mult * risk_profile.target_scale * self.atr) if atr_target_mult else None
        self.trail_dist = (trailing_atr_mult * risk_profile.trail_scale * self.atr) if trailing_atr_mult else None
        self.time_limit = int(time_stop_bars * risk_profile.time_scale) if time_stop_bars else None

        # Tracking
        self.bars_held = 0
        self.best_price = entry_price
        self.worst_price = entry_price
        self._trailing_stop: Optional[float] = None

        if self.trail_dist:
            if direction == "long":
                self._trailing_stop = entry_price - self.trail_dist
            else:
                self._trailing_stop = entry_price + self.trail_dist

    def check(self, high: float, low: float, close: float) -> Optional[str]:
        """Check if any exit condition is met for this bar.

        Args:
            high: Current bar high
            low: Current bar low
            close: Current bar close

        Returns:
            Exit reason string, or None if no exit triggered.
        """
        self.bars_held += 1

        # Update MFE/MAE tracking
        if self.direction == "long":
            self.best_price = max(self.best_price, high)
            self.worst_price = min(self.worst_price, low)
        else:
            self.best_price = min(self.best_price, low)
            self.worst_price = max(self.worst_price, high)

        # 1. Fixed stop loss
        if self.stop_dist is not None:
            if self.direction == "long":
                stop_level = self.entry_price - self.stop_dist
                if low <= stop_level:
                    return "stop_loss"
            else:
                stop_level = self.entry_price + self.stop_dist
                if high >= stop_level:
                    return "stop_loss"

        # 2. Fixed target
        if self.target_dist is not None:
            if self.direction == "long":
                target_level = self.entry_price + self.target_dist
                if high >= target_level:
                    return "target"
            else:
                target_level = self.entry_price - self.target_dist
                if low <= target_level:
                    return "target"

        # 3. Trailing stop
        if self.trail_dist is not None and self._trailing_stop is not None:
            if self.direction == "long":
                new_trail = high - self.trail_dist
                self._trailing_stop = max(self._trailing_stop, new_trail)
                if low <= self._trailing_stop:
                    return "trailing_stop"
            else:
                new_trail = low + self.trail_dist
                self._trailing_stop = min(self._trailing_stop, new_trail)
                if high >= self._trailing_stop:
                    return "trailing_stop"

        # 4. Time stop
        if self.time_limit is not None and self.bars_held >= self.time_limit:
            return "time_stop"

        return None

    @property
    def max_drawdown_pct(self) -> float:
        """Maximum adverse excursion as a percentage of entry price."""
        if self.direction == "long":
            return (self.entry_price - self.worst_price) / self.entry_price
        else:
            return (self.worst_price - self.entry_price) / self.entry_price

    @property
    def max_profit_pct(self) -> float:
        """Maximum favorable excursion as a percentage of entry price."""
        if self.direction == "long":
            return (self.best_price - self.entry_price) / self.entry_price
        else:
            return (self.entry_price - self.best_price) / self.entry_price

File: go-strats/pkg/exits/exits.go
// Package exits implements exit management for the probe engine.
//
// Handles ATR-based stops, targets, trailing stops, and time stops
// with risk-profile scaling. Mirrors Python src/strats_prob/exits.py.
package exits

import (
	"math"

	"github.com/algomatic/strats100/go-strats/pkg/types"
)

// ExitManager manages exit logic for a single trade.
// Initialized at entry with risk-profile-scaled parameters.
// Call Check() each bar to determine if the trade should be closed.
type ExitManager struct {
	EntryPrice  float64
	Direction   types.Direction
	ATR         float64 // ATR at entry
	RiskProfile types.RiskProfile

	// Scaled distances (0 means not used)
	StopDist   float64
	TargetDist float64
	TrailDist  float64
	TimeLimit  int // 0 means not used

	// Tracking
	BarsHeld     int
	BestPrice    float64
	WorstPrice   float64
	trailingStop float64
	hasTrail     bool
	barPnLs      []float64
}

// NewExitManager creates a new ExitManager for a trade entry.
func NewExitManager(
	entryPrice float64,
	direction types.Direction,
	atrAtEntry float64,
	strategy *types.StrategyDef,
	riskProfile types.RiskProfile,
) *ExitManager {
	em := &ExitManager{
		EntryPrice:  entryPrice,
		Direction:   direction,
		ATR:         atrAtEntry,
		RiskProfile: riskProfile,
		BestPrice:   entryPrice,
		WorstPrice:  entryPrice,
		barPnLs:     make([]float64, 0, 64),
	}

	// Scale multipliers by risk profile
	if strategy.ATRStopMult > 0 {
		em.StopDist = strategy.ATRStopMult * riskProfile.StopScale * atrAtEntry
	}
	if strategy.ATRTargetMult > 0 {
		em.TargetDist = strategy.ATRTargetMult * riskProfile.TargetScale * atrAtEntry
	}
	if strategy.TrailingATRMult > 0 {
		em.TrailDist = strategy.TrailingATRMult * riskProfile.TrailScale * atrAtEntry
		em.hasTrail = true
		if direction == types.Long {
			em.trailingStop = entryPrice - em.TrailDist
		} else {
			em.trailingStop = entryPrice + em.TrailDist
		}
	}
	if strategy.TimeStopBars > 0 {
		em.TimeLimit = int(float64(strategy.TimeStopBars) * riskProfile.TimeScale)
	}

	return em
}

// Check evaluates whether any exit condition is met for the current bar.
// Returns the exit reason string, or empty string if no exit triggered.
func (em *ExitManager) Check(high, low, closePrice float64) string {
	em.BarsHeld++

	// Update MFE/MAE tracking
	if em.Direction == types.Long {
		em.BestPrice = math.Max(em.BestPrice, high)
		em.WorstPrice = math.Min(em.WorstPrice, low)
	} else {
		em.BestPrice = math.Min(em.BestPrice, low)
		em.WorstPrice = math.Max(em.WorstPrice, high)
	}

	// Track bar-by-bar P&L % for pnl_std computation
	var barPnL float64
	if em.Direction == types.Long {
		barPnL = (closePrice - em.EntryPrice) / em.EntryPrice
	} else {
		barPnL = (em.EntryPrice - closePrice) / em.EntryPrice
	}
	em.barPnLs = append(em.barPnLs, barPnL)

	// 1. Fixed stop loss
	if em.StopDist > 0 {
		if em.Direction == types.Long {
			stopLevel := em.EntryPrice - em.StopDist
			if low <= stopLevel {
				return "stop_loss"
			}
		} else {
			stopLevel := em.EntryPrice + em.StopDist
			if high >= stopLevel {
				return "stop_loss"
			}
		}
	}

	// 2. Fixed target
	if em.TargetDist > 0 {
		if em.Direction == types.Long {
			targetLevel := em.EntryPrice + em.TargetDist
			if high >= targetLevel {
				return "target"
			}
		} else {
			targetLevel := em.EntryPrice - em.TargetDist
			if low <= targetLevel {
				return "target"
			}
		}
	}

	// 3. Trailing stop
	if em.hasTrail {
		if em.Direction == types.Long {
			newTrail := high - em.TrailDist
			if newTrail > em.trailingStop {
				em.trailingStop = newTrail
			}
			if low <= em.trailingStop {
				return "trailing_stop"
			}
		} else {
			newTrail := low + em.TrailDist
			if newTrail < em.trailingStop {
				em.trailingStop = newTrail
			}
			if high >= em.trailingStop {
				return "trailing_stop"
			}
		}
	}

	// 4. Time stop
	if em.TimeLimit > 0 && em.BarsHeld >= em.TimeLimit {
		return "time_stop"
	}

	return ""
}

// MaxDrawdownPct returns the maximum adverse excursion as a percentage of entry price.
func (em *ExitManager) MaxDrawdownPct() float64 {
	if em.Direction == types.Long {
		return (em.EntryPrice - em.WorstPrice) / em.EntryPrice
	}
	return (em.WorstPrice - em.EntryPrice) / em.EntryPrice
}

// MaxProfitPct returns the maximum favorable excursion as a percentage of entry price.
func (em *ExitManager) MaxProfitPct() float64 {
	if em.Direction == types.Long {
		return (em.BestPrice - em.EntryPrice) / em.EntryPrice
	}
	return (em.EntryPrice - em.BestPrice) / em.EntryPrice
}

// PnLStd returns the standard deviation of bar-by-bar P&L % during the hold period.
func (em *ExitManager) PnLStd() float64 {
	if len(em.barPnLs) < 2 {
		return 0.0
	}
	// Compute mean
	sum := 0.0
	for _, v := range em.barPnLs {
		sum += v
	}
	mean := sum / float64(len(em.barPnLs))
	// Compute variance (population std, ddof=0)
	varSum := 0.0
	for _, v := range em.barPnLs {
		diff := v - mean
		varSum += diff * diff
	}
	return math.Sqrt(varSum / float64(len(em.barPnLs)))
}

// TrailingStopLevel returns the current trailing stop level.
func (em *ExitManager) TrailingStopLevel() float64 {
	return em.trailingStop
}

// GetExitPrice determines the exit price based on the exit reason.
func GetExitPrice(
	exitReason string,
	direction types.Direction,
	entryPrice float64,
	em *ExitManager,
	closePrice float64,
) float64 {
	switch exitReason {
	case "stop_loss":
		if em.StopDist > 0 {
			if direction == types.Long {
				return entryPrice - em.StopDist
			}
			return entryPrice + em.StopDist
		}
	case "target":
		if em.TargetDist > 0 {
			if direction == types.Long {
				return entryPrice + em.TargetDist
			}
			return entryPrice - em.TargetDist
		}
	case "trailing_stop":
		if em.hasTrail {
			return em.trailingStop
		}
	}
	// Signal exit, time stop, or unknown: exit at close
	return closePrice
}
